dbms/src/Formats/CSVRowInputStream.cpp

```diff
@@ -191,46 +162,104 @@ void CSVRowInputStream::readPrefix()
             while (checkChar(format_settings.csv.delimiter, istr));
 
             skipDelimiter(istr, format_settings.csv.delimiter, true);
+
+            for (size_t column = 0; column < read_columns.size(); column++)
+            {
+                if (!read_columns[column])
+                {
+                    have_always_default_columns = true;
+                    break;
+                }
+            }
+
+            return;
         }
         else
         {
-            setupAllColumnsByTableSchema();
-            skipRow(istr, format_settings.csv, column_indexes_for_input_fields.size());
+            skipRow(istr, format_settings.csv, header.columns());
         }
     }
-    else
+
+    /// The default: map each column of the file to the column of the table with
+    /// the same index.
+    read_columns.assign(header.columns(), true);
+    column_indexes_for_input_fields.resize(header.columns());
+
+    for (size_t i = 0; i < column_indexes_for_input_fields.size(); ++i)
     {
-        setupAllColumnsByTableSchema();
+        column_indexes_for_input_fields[i] = i;
     }
 }
 
-
+/** If you change this function, don't forget to change its counterpart
+  * with extended error reporting: parseRowAndPrintDiagnosticInfo().
+  */
 bool CSVRowInputStream::read(MutableColumns & columns, RowReadExtension & ext)
 {
     if (istr.eof())
         return false;
 
     updateDiagnosticInfo();
 
-    String tmp;
+    /// Track whether we have to fill any columns in this row with default
+    /// values. If not, we return an empty column mask to the caller, so that
+    /// it doesn't have to check it.
+    bool have_default_columns = have_always_default_columns;
+
+    const auto delimiter = format_settings.csv.delimiter;
     for (size_t input_position = 0; input_position < column_indexes_for_input_fields.size(); ++input_position)
     {
         const auto & column_index = column_indexes_for_input_fields[input_position];
         if (column_index)
         {
-            skipWhitespacesAndTabs(istr);
-            data_types[*column_index]->deserializeAsTextCSV(*columns[*column_index], istr, format_settings);
-            skipWhitespacesAndTabs(istr);
+            const auto & type = data_types[*column_index];
+            if (*istr.position() == delimiter
+                    && format_settings.csv.empty_as_default)
```

Comment:
Do we check for empty values at the last column?

```
INSERT INTO table (a, b) FORMAT CSV
Hello,
```

Reply:
(it requires comparison with \r and \n; we assume that unquoted newlines are not possible in value)